Mathlib/CategoryTheory/Sites/IsConstant.lean

```diff
@@ -0,0 +1,204 @@
+/-
+Copyright (c) 2024 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.CategoryTheory.Adjunction.FullyFaithful
+import Mathlib.CategoryTheory.Sites.ConstantSheaf
+import Mathlib.CategoryTheory.Sites.DenseSubsite
+import Mathlib.CategoryTheory.Sites.PreservesSheafification
+/-!
+
+# Constant objects in sheaf categories.
+
+This file defines the notion `Sheaf.IsConstant`. A sheaf `F` *is constant* if the counit
+`(F(*))^cst ⟶ F` of the constant sheaf adjunction is an isomorphism at `F`. Here `*` denotes a
+particular chosen terminal object of the defining site, and `cst` denotes the constant sheaf.
+
+It is convenient to take an arbitrary terminal object; one might want to use this construction to
+talk about sheaves being constant when the defining site has a particularly convenient terminal
+object, such as the one element space in `CompHaus`.
+
+## Main results
+
+* `isConstant_iff_mem_essImage` : A sheaf is constant if and only if it is in the essential image
+of the constant sheaf functor, if the constant sheaf functor is fully faithful.
+* `isConstant_iff_of_equivalence` : The property of a sheaf of being constant is invariant under
+equivalence of sheaf categories, if the constant sheaf functor is fully faithful (the assumption
+on the constant sheaf functor is probably unnecessary, but it's low priority to prove this in more
+generality, because in all cases of interest, the assumption holds).
+* `isConstant_iff_forget` : Given a "forgetful" functor `U : A ⥤ B` a sheaf `F : Sheaf J A` is
+constant if and only if the sheaf given by postcomposition with `U` is constant.
+
+## Future work
+
+* (Dagur) Use `isConstant_iff_forget` to prove that a condensed module is discrete if and only if
+its underlying condensed set is discrete.
+-/
+
+open CategoryTheory Limits Functor Adjunction Opposite Category Functor
+
+namespace CategoryTheory.Sheaf
+
+variable {C : Type*} [Category C] (J : GrothendieckTopology C) {A : Type*} [Category A]
+  [HasWeakSheafify J A] {t : C} (ht : IsTerminal t)
+
+/--
+A sheaf is constant if the counit of the constant sheaf adjunction applied to it is an isomorphism.
+-/
+abbrev IsConstant (F : Sheaf J A) : Prop := IsIso ((constantSheafAdj J A ht).counit.app F)
```

Comment:
It would be better to ask that `F` belongs to the essential image of the functor `constantSheaf` (note also that it would not depend on the existence/a choice of a terminal object.). (If `C` corresponds to a topological space that is not connected, the definition `IsIso (....counit.app...)` is not the right definition of `IsConstant`.) Changing the definition should not be a problem for your applications to condensed objects.